### ISSUE TITLE
Conda distribution

### ISSUE THREAD

**Issue by GiovanniBussi** [plumed/plumed2#412]
@tonigi mentioned the possibility to distribute PLUMED binaries through conda (see #411). This would amount basically to distributing the driver, the library, and the python wrappers.

[Here]( https://anaconda.org/acellera/plumed2) we would have a good starting point.

I am personally not familiar with conda (the only time I tried it I made a mess), but I think it would be worth having a look at what can be done.

I open this issue in case anyone wants to comment on this topic.

**Comment by tonigi** [comment]
IMHO, Conda provides a "no hassle" and relatively clean way to install binaries (even non-python ones) on the three major platforms at the cost of yet another ecosystem to maintain.  

I guess the "real" users of Plumed are power users who need to compile themselves, so Conda can't reasonably be the "main" distribution method. 

On the plus side, `conda-forge` provides automated CI builds, and it has been adopted by several MD-related projects (notably OpenMM and all major analysis libraries).  I'm attempting [here](https://github.com/conda-forge/staged-recipes) a very preliminary conda-forge recipe (only working on linux for now... will keep trying for a couple of days).

**Comment by tonigi** [comment]
Just an update: the conda-forge packaging seems to work (links below). Overall conda-forge required a surprising amount of tweaking to make it work (it runs three different CI systems). So, I am not sure whether it's worth to "promote" it, i.e. to ask conda-forge maintainers to make it into an auto-building "feedstock". The step would start building and distributing the plumed2 conda packages in conda-forge, but it is somewhat irreversible.

* "Staged recipe" (i.e. build only, not distributing) https://github.com/tonigi/staged-recipes
* Thread with CIs https://github.com/conda-forge/staged-recipes/pull/7092

**Comment by GiovanniBussi** [comment]
Thanks @tonigi !!!!

I double checked the build and everything seems fine. I think it is working on Mac (travis-ci) and Linux (circle-ci). Not sure Windows is also built.

I have some small comments on the python stuff that I put at the end of this post.

About the decision of whether uploading from travis or from conda-forge I don't know which is the best way. If I understand correctly, the good thing of conda-forge is that they take care of having their environment working on travis-ci/circle-ci. For example, **they** will take care of what will happen in January when the used xcode will be dismissed (there is a big warning in the travis build). On the other hand, the bad thing is that **we** will have to make sure that plumed continues to build correctly using the protocol that you patiently prepared.

Are there other consequences?

A maybe stupid question: in case we have a new release and we manage conda independently, I guess we should setup our travis-ci to upload the new version. If we use conda-forge should we just push a version tag somewhere?

About irreversibility: if at some point we get bored of maintaining the compatibility, will the conda-forge just stop update after a given version? (this would not be a problem I think).


## Note on Python interface
I have a micro comment on the way you install the python wrappers. If I understand correctly, on all the machines you use something like
````
cd python
make pip
pip install
````

If that's the case:
1. I would disable python in configure (with `--disable-python`).
2. You do not need the patch `0001-use-python-kernel-in-prefix.patch`. Indeed, that part of the Makefile is only needed if you install the wrappers inside PLUMED.
3. The correct way to do what I think you want to do is to set env var `plumed_default_kernel` before running pip. Something like:
````
cd python
make pip
export plumed_default_kernel=/path/to/plumed/kernel
pip install
````
Here I don't know how to guess the extension and the path correctly in the conda environment.

**Comment by tonigi** [comment]
Hi @GiovanniBussi !  Here’s some tentative answers (as far as my knowledge goes).

The analysis is correct. A conda forge “feedstock” build is triggered editing `meta.yml` in the corresponding repository (say, one changes the version tag or the download address). If all goes well, packages automatically end up in conda-forge channel. This is handled by a (large) number of support scripts which are maintained by conda-forge, which also means that the build may break at any time (due to unexpected changes either to the scripts or to conda).  “Abandoned” packages will stay at their latest version, presumably spamming the maintainer.

On the other hand, self-making a conda build seems to have the following potential consequences:
* possibly lesser-quality builds: in principle one should build with conda-only libraries and compilers, which is somewhat enforced in circle-ci 
* one will upload to a “plumed” conda channel, not conda-forge (no big deal)
* if abandoned, no spamming is expected  :) 
* uploading from travis should be easy (“conda upload” with a token) anyway, especially given that you already have a great travis workflow 

Indeed I disabled the windows build (as most packages do) because it required further fiddling. Regarding the python interface, there should be no problem to change.

**Comment by tonigi** [comment]
Here's an attempt at a self-made conda build on travis [link](https://github.com/tonigi/plumed2/tree/v2.5-conda). Note that the default conda-build configuration uses its own gcc-7.3.0 (with openmp). And [here is travis](https://travis-ci.org/tonigi/plumed2)

Probably mkl and mpi can also be activated if desired. Omnia, from what I understand, is obsolete.

**Comment by GiovanniBussi** [comment]
As far as I understand maintaining conda-forge would be more difficult than having our own channel... So I am ok with trying our channel

**Comment by hainm** [comment]
> ambertools: dedicated channel AmberMD plus acpype

@GiovanniBussi: Just FYI, amber's hosting its channel in its own server: http://ambermd.org/downloads/ambertools/conda/

**Comment by dwhswenson** [comment]
Came across this issue while thinking I should suggest to @gtribello that PLUMED should make itself conda-installable -- glad to hear you're already working on it! This will help us move the PLUMED CVs for OpenPathSampling package into the core of OPS, because it will mean that installing PLUMED becomes much easier in our testing process.

A couple comments:

### Channels

> * mdtraj : mostly omnia, then conda-forge and acellera

All the omnia projects, including MDTraj, are [moving away from the omnia channel and to conda-forge](http://www.omnia.md/blog/2017/5/18/omnias-conda-distribution-now-linked-with-conda-forge). MDTraj is at 1.9.2 on conda-forge; 1.9.1 on omnia. A few stragglers are left behind in the omnia channel, mainly because OpenMM can't yet be built on conda-forge, so that also leaves behind any packages that include OpenMM as a requirement. (OPS lives there because OpenMM is technically a requirement -- for now.) 

I'd recommend reading over the linked Omnia blog post as a warning on the hazards of running your own channel. Their problems probably won't apply to you, but worth you knowing why they switched away from that approach.

### Binary compatibility

[Conda-forge packages built with the `gcc7` label are compatible with conda defaults](https://github.com/conda-forge/conda-forge.github.io/issues/616#issuecomment-449074631) (later comment from an issue mentioned earlier). In fact, I'm not quite clear on the details, but I think the `gcc7` label behavior became the standard conda-forge behavior in January (and actually uses GCC 8). Can't find this in docs, but they did say it [on Twitter](https://twitter.com/condaforge/status/1083379357840744448).

---

In any case, I'm looking forward to `conda install -c <some-channel> plumed`!

**Comment by GiovanniBussi** [comment]
@dwhswenson thanks for your feedback. Branch #416 is a work-in-progress but essentially functional (although I think I made some mess trying to install xdrfile library in the last commit).

I am personally still a bit skeptical about the whole conda approach, and I would rather invest in making installation from source as easy as possible. I think having MPI parallelism would be pretty complicated within conda (unless you force users to a specific MPI version). I am already impressed by the fact that you need to choose the python version first, that for PLUMED is a bit nonsensical given that the python wrappers are a thin layer on top of a large binary library. I mean: having a separate binary to be used if I call plumed from python 3.6 or 3.7 is a bit strange to me. Once I looked at the openMM binary files and, since they support 3-4 CUDA versions, 3 python versions, and three operating systems, the number of packages grows combinatorially...

Anyway, I see that there is a category of users for which standard settings would be fine, so perhaps it is worth to make the effort.

**Comment by tonigi** [comment]
I can't speak for the "official" conda plumed release (if any), but the test build on my channel could be representative of what to expect:

```
 conda install -c tonigi/label/testing plumed2 
```

Indeed I agree with @GiovanniBussi .   Yet, it's probably still worthwhile to provide  (only?) a basic unoptimized package.

**Comment by dwhswenson** [comment]
@tonigi : Thanks for this. I meant to write last week -- I decided to try use it myself during Gareth's tutorial. In addition, it turned out that a few participants hadn't installed PLUMED, and had trouble the day of the tutorial, so we had them use the conda install as well (they already had conda installed for other things).

I also agree that the best thing for most users is to compile directly, and I don't think that the PLUMED build process is particularly difficult. The main advantages that conda provides are: (1) other codes that have PLUMED integrations appreciate using the pre-built version for regression tests of the integration; (2) quick setup during tutorials for that inevitable person who has a weird compiler setup and has trouble building -- conda gives the option of a clean and controlled environment.

Thanks again for providing it!

**Comment by GiovanniBussi** [comment]
I close this since we now have a basic workflow to install plumed from conda. Builds will be available officially starting with v2.5.2 from [this channel](https://anaconda.org/plumed/plumed).

**Comment by GiovanniBussi** [comment]
Regarding conda distribution, I managed to split the [plumed package](https://anaconda.org/plumed/plumed) and the [python interface](https://anaconda.org/plumed/py-plumed). This means that we can easily support multiple python versions without recompiling the full library. I am using the same package names as in MacPorts for consistency, that is, this installs the plumed library:
````
conda install -c plumed/label/testing plumed
````
and this installs the python wrappers:
````
conda install -c plumed/label/testing py-plumed
````
Currently the python wrappers are compiled for python 2.7, 3.6 and 3.7 (same as in MacPorts)

**Comment by GiovanniBussi** [comment]
An update on this topic. After some experiment, I realized that distributing through [conda-forge](https://conda-forge.org) will be easier in the long run. The mechanism is similar to MacPorts: it is a curated set of packages, contributed by the community. The team is very helpful and responds very quickly. The packages we depend on are up to date and I already managed to contribute [xdrfile](https://anaconda.org/conda-forge/xdrfile). There is a gromacs package on bioconda, and bioconda relies on conda-forge, so in principle we could even try to add plumed to that gromacs version.

Thus, I decided contribute a [plumed package there](https://anaconda.org/conda-forge/plumed). It can be installed with
````
conda install -c conda-forge plumed
````
(currently this is v2.5.1).

In the next days I will take care of python wrappers as well. Finally, I think we can keep our channel for testing versions and/or nightly builds, or if we want to distribute pre-patched MD codes. I will also update the manual accordingly.
